enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,535 @@
+
+Let's consider an operator that installs a database. Assume that the operator already communicates that a database migration is in progress by setting the "Migrating" condition type in the status of the `foo` CustomResourceDefinition and the operator does not want to upgrade when a database is Migrating. In this case, the operator's CRD should include the following annotations:
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata: {
+  name: foo
+  annotations: {
+    "olm.Condition.Upgradeable" : "Migrating"
+    "olm.Condition.Important" : "BadConnectivity,UnhealthyDatabase"
+  }
+}
+```
+
+When this annotation is present on a CRD, OLM will append the following `Probe` CR to the `installPlan`:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: Probe
+metadata:
+  name: foo-operator.v1.0.0
+  namespace: operators
+spec:
+  specOverride: crdAnnotations # Changes to the CRD's annotations should override the spec. SPO will watch the CRD for changes to the "OLM Supported Condition" annotations and update the spec accordingly.
+  probeResources:
+  - name: foo.example.com 
+    upgradeable: # Converted to array
+    - Migrating
+    important: # Converted to array
+    - BadConnectivity
+    - UnhealthyDatabase
+```
+
+When the`probe` resource above is created, OLM will spin up an SPO pod which collects the information from conditions of `foo.example.com` CRs and communicate it to OLM via the `probe`'s status accordingly. The permissions of the probe will match the same permissions of the operator.
+
+>Note: If the probe's spec is nil, SPO will not create a pod and the operator will be expected to update the CRs status itself.
+
+In this example, the pod will be responsible for maintaining the following fields in the status of the `probe` resource:
+
+```yaml
+status:
+  upgradeable: # A boolean that represents if the operator may be upgraded - aggregated from all probeResources below.
+  probeResources: # The list of resources the probe will watch and the conditions of those resources.
+```
+
+Consider an upgrade where the a single `Foo` CR exists on cluster:
+
+```yaml
+apiVersion: foo.example.com/v1
+kind: Foo
+metadata:
+  name: foo-example
+  namespace: default"
+spec:
+  ...
+  ...
+  ...
+status:
+  probeResources:
+  - resource: "foo.example.com"
+    - uid: "xxx"
+      conditions:
+      - type: Migrating # Condition 1
+        status: true
+        reason: "CriticalProcess" 
+        message: "Migrating old database schema to new format"
+      - type: unhealthyDatabase # Condition 2
+        status: true
+        reason: "LivenessProbeFailure"
+        message: "The foo-database failed its liveness probe."
+      - type: badConnectivity # Condition 3
+        status: true
+        reason: "404" 
+        message: "Unable to reach the database"
+```
+
+From this CR, OLM will update the Probe CR like so:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: Probe
+metadata:
```

Comment:
> To be aggregated by an Operator resource, a Probe will need to be "adopted". Adoption should consist of finding the package and install namespace that originated this Probe, and labelling it with the operator name (which is a combination of package and install namespace).

Agreed.

> This also brings up an interesting question, can a single probe be owned by multiple operators? 

I do not believe that a single probe should be owned by multiple operators, each probe should represent the state of a single operator.

> What does multiple single namespace installs for a CRD look like.
 
This is a great point and one that I have to update the PR to point out. The probe should always be in the same namespace as the operator that owns it. This should make it fairly easy to capture which probe represents which operator.

Reply:
It's hard for me to think of a -- nice -- set of rules we can apply to figure out which CRs to pull into a Probe then.

For namespaced-scoped CRs, you could say only the CRs in the install namespace are included, but that limits the author in the case where a CR in another namespace is critical to its operation. Further, for cluster-scoped CRs, that heuristic is no longer useful and now we need to make a decision about how they are shared amongst probes that reference them.

This seems to be another case where our scoping discussion rears its head again.

... but since this needs to be compatible with our current design, I wonder if we can sidestep the issue by placing the annotations on individual CRs, pushing off the "all CRs in a tenant" use case for now. The implication is that authors now need to include a `Probe` resource in their manifests, but I'm not sure how different that is from adding annotations to an existing CRD manifests in the first place.

Reply:
> For namespaced-scoped CRs, you could say only the CRs in the install namespace are included, but that limits the author in the case where a CR in another namespace is critical to its operation. Further, for cluster-scoped CRs, that heuristic is no longer useful and now we need to make a decision about how they are shared amongst probes that reference them.

The probe will look for CRs in all namespaces included in the operator group that the operator is installed in. I do not see where there would be an issue identifying which CRs are pulled into a probe.

Reply:
> Further, for cluster-scoped CRs, that heuristic is no longer useful and now we need to make a decision about how they are shared amongst probes that reference them.

If the operator is installed at a cluster level, OLM would not allow a new operator providing the same API to be installed so I don't believe that this applies.